Allow scoped allocation

Marwes opened the pull request:
I gave implementing scoped allocation a shot since I had an idea about how allow Recycled objects with both & and Rc without duplicating any code. I think it turned other rather good but I wanted to get some feedback before proceeding. One nice thing of this implementation is how the Rc based pool can allocate both Recycled and ScopedRecycled.

The code will still need some cleanup, some tests and benchmarks added, and I think it may be a good idea to make ScopedPool etc newtypes as well to avoid their internal implementation leaking out.

Marwes committed:
Allow scoped allocation by parametrization over the pool type

Recycled is split into two types, RcRecycled and Recycled. RcRecycled has the old behaviour and Recycled holds the pool only by a & reference allowing faster allocation/deallocation since the Rc does not have to be incermented/decremented.

zslayton commented:
This is great! Some thoughts:
- I love that you've generalized what kind of borrowed value is being used. This makes me wonder if we could easily add `Arc` to the mix `where T : Send` and get values that could be reclaimed across threads.
- It took me a minute to realize what was happening in `test_attach_scoped_from_rc()`. Being able to have the `Rc` pool optionally work with `ScopedRecycled` is genius.
- I'm excited that all of this flexibility / functionality didn't require the use of `unsafe`.
- I'd like to tweak some of the method names (e.g. I'd prefer that `with_size_scoped(10)` be `scoped_with_size(10)`). I'm also curious whether there might be a more descriptive name for `GenRecycled`, but nothing's springing to mind at the moment.

Thanks very much for taking the time to work on this!

zslayton commented on benches/lib.rs:
```diff
@@ -41,6 +41,19 @@ mod tests {
     });
   }
+  #[bench]
+  fn bench02_pooled_scoped_allocation_speed(b: &mut Bencher) {
```
I started numbering the benchmark names so that related test results would be shown side-by-side. (e.g. `bench01_standard_allocation_speed` and `bench02_pooled_allocation_speed`.) It works, but in situations like this where you'd like to add another test, it requires re-numbering all of the functions. Any suggestions for a better approach?

Marwes replied:
Maybe it would work to write the name in prefix form instead of numbering the benchmarks? Assuming they are just ordered lexically this should work:

```
  fn allocation_speed_standard(b: &mut Bencher)
  fn allocation_speed_pooled(b: &mut Bencher)
  fn allocation_speed_pooled_scoped(b: &mut Bencher)
```

zslayton replied:
Works for me!

Marwes commented:
- Arc is a good idea but it would require a Mutex instead of a RefCell so it might be a bit more work to add it since I currently require the bound `AsRef<RefCell<Vec<T>>>`.
- :)
- :)
- I can do that, I just took the first name I thought of to make it easy to search and replace any discrepancies.

zslayton commented:
Thought about this a bit more this afternoon: I think the simplest function names should produce `ScopedRecycled` since it has the least overhead. Then we can name the functions producing the `Rc` flavor of things `new_rc`, `new_rc_from`, etc. Whaddya think?

Marwes commented:
I changed the type definitions to newtypes which increased the code size quite a bit since I needed to forward the functions on Pool/Recycled. I added a macro to do this so at least it only has to be done once instead of for every Pool/Recycled type.

I also renamed the scoped variations to take the rc names and the rc variations to have a rc suffix/prefix. Let me know what you think

If you'd like I can rebase this PR and squash some commits together.

zslayton commented:
Cool! I think squashing the commits is a good idea. Last thing I want to consider before merging: would it be better if there were only one `Pool` type? `RcPool` can issue both `Recycled` and `RcRecycled` values and as far as I can see, the user would only take the `RcRecycled` performance hit if (s)he used the `*_rc` methods. Is that accurate? It would simplify the API a little bit.

Marwes commented:
It would simplify the API a bit so that might be good idea. That being said having it inside an Rc involves an additional pointer dereference on each allocation so it is not just overhead on construction/destruction of the pool. I ran a benchmark on my computer and got these results.

```
test tests::bench02_pooled_allocation_speed_rc            ... bench:    702868 ns/iter (+/- 129790)
test tests::bench02_pooled_allocation_speed_rc_scoped     ... bench:    543640 ns/iter (+/- 55838)
test tests::bench02_pooled_allocation_speed_scoped        ... bench:    512913 ns/iter (+/- 51501)
```

Its just a 5% difference so I guess it isn't really worth keeping.

Marwes commented:
Replaced Pool with RcPool and squashed.

zslayton commented:
Ah, thanks for running the benchmark. It does seem to be a pretty small difference. I'd say let's promote `RcPool` to being the only kind of `Pool` for now. It may be possible to store a reference to the un-`Rc`'ed `RefCell` using `unsafe` if we need to get that extra speed boost. I'll have to do some research.

zslayton commented:
Excellent! Thanks again!
